File: py-usfm-parser/tests/test_auto_fix.py
import pytest
import glob
import json
from src.usfm_grammar import Validator

test_files = glob.glob("../tests/autofix/*")
sample_usj_files = glob.glob("../tests/specExamples/*/origin.json")

@pytest.mark.parametrize('file_path', test_files)
def test_auto_fix_erros(file_path):
    '''Tests if input parses with or without errors, as expected'''
    with open(file_path, 'r', encoding='utf-8') as usfm_file:
        usfm_string = usfm_file.read()

    checker = Validator()
    assert type(checker.is_valid_usfm(usfm_string)) == bool
    fixed_usfm = checker.auto_fix_usfm(usfm_string)
    assert checker.is_valid_usfm(fixed_usfm) == True, checker.message


@pytest.mark.parametrize('file_path', sample_usj_files)
def test_validate(file_path):
    '''Tests if input parses with or without errors, as expected'''
    with open(file_path, 'r', encoding='utf-8') as usfm_file:
        usj_string = usfm_file.read()


    checker = Validator()
    correct_usj = json.loads(usj_string)
    assert checker.is_valid_usj(correct_usj) == True

    incorrect_usj = json.loads(usj_string.replace('code', 'coooode').replace('content', 'contents'))
    assert checker.is_valid_usj(incorrect_usj) == False

File: node-usfm-parser/test/test_auto_fix.js
const assert = require('assert');
const fs = require('node:fs');
const {glob} = require('glob');
const {Validator} = require("../src/index");

const TEST_DIR = "../tests";
const allUSFMFiles = glob.sync(TEST_DIR+'/autofix/*');
const sampleUSJs = glob.sync(TEST_DIR+'/specExamples/*/origin.json')

describe("Try autofixing errors in USFM", () => {

  	allUSFMFiles.forEach(function(value) {
      it(`Fix ${value}`, (inputUsfmPath=value) => {
        //Tests if input parses without errors
        const testVaidator = new Validator()
        assert(testVaidator instanceof Validator)
        const inputUsfm = fs.readFileSync(inputUsfmPath, 'utf8')
        const firstTest = testVaidator.isValidUSFM(inputUsfm);
        const fixedUsfm = testVaidator.autoFixUSFM(inputUsfm);
        const secondTest = testVaidator.isValidUSFM(fixedUsfm);
        // assert.ok(!firstTest);
        assert.ok(secondTest);
      });
    });

});

describe("Validate USJ", () => {
    sampleUSJs.forEach(function(value) {
      it(`Validate ${value}`, (inputUsjPath=value) => {
        //Tests if input parses without errors
        const testVaidator = new Validator()
        assert(testVaidator instanceof Validator)
        const inputUsj = fs.readFileSync(inputUsjPath, 'utf8')
        const usj = JSON.parse(inputUsj);
        // assert.ok(!firstTest);
        assert.ok(testVaidator.isValidUSJ(usj));
      });
    });

    sampleUSJs.forEach(function(value) {
      it(`Validate ${value} and report error`, (inputUsjPath=value) => {
        //Tests if input parses without errors
        const testVaidator = new Validator()
        assert(testVaidator instanceof Validator)
        let inputUsj = fs.readFileSync(inputUsjPath, 'utf8')
        inputUsj = inputUsj.replace("code", "cooode");
        inputUsj = inputUsj.replace("content", "contents");
        const usj = JSON.parse(inputUsj);
        assert.ok(!testVaidator.isValidUSJ(usj));
        assert(testVaidator.message !== "");
      });
    });
});
